core/src/main/java/software/amazon/awssdk/client/BaseClientHandler.java

```diff
@@ -54,11 +56,16 @@ ExecutionContext createExecutionContext(RequestConfig requestConfig) {
                 : clientConfiguration.credentialsProvider();
 
         ClientOverrideConfiguration overrideConfiguration = clientConfiguration.overrideConfiguration();
+
+        AwsCredentials credentials = credentialsProvider.getCredentials();
+
+        Validate.validState(credentials != null, "Credentials can't be null.");
```

Comment:
Maybe "credential provider must never return null" to indicate that it's the credential provider's fault (which presumably they've overridden because we'd never have a bug that causes us to return null!)